### PR TITLE
Support calling and creating AGGREGATEs

### PR DESCRIPTION
**Pull request by xuebinsu**
For data scientists, aggregates are essential for retrieving summaries
of the data. This patch adds new functions for calling existing
aggregates in Greenplum, and create new aggregates from Python
functions.

Note that this patch only includes most fundamental support for
aggregates. Many features, e.g. final functions, combine functions,
and composite state type, are NOT supported yet.

### CODE REVIEW

**Review by Quasimewdo — COMMENTED** [review]
LGTM

**Review by Quasimewdo — APPROVED** [review]

File: greenplumpython/func.py [hunk 1]
```diff
@@ -49,23 +99,24 @@ def make_function_call(
...
+                f"{param.name} {primitive_type_map[param.annotation]}"
+                for param in func_sig.parameters.values()
             ]
         )
         # FIXME: include things in func.__closure__
         func_lines = textwrap.dedent(inspect.getsource(func)).split("\n")
         func_body = "\n".join([line for line in func_lines if re.match(r"^\s", line)])
         if db is None:
             for arg in args:
+                print(arg.db)
```
Comment (beeender) [inline greenplumpython/func.py:119]: don't forget to remove this.

**Review by beeender — COMMENTED** [review]
will do another round of interview after the conflicts get resolved.

File: tests/test_func.py [hunk 2]
```diff
@@ -80,3 +80,51 @@ def multiply(a: int, b: int) -> int:
...
+    @gp.create_aggregate
+    def my_sum(result: int, val: int) -> int:
+        if result is None:
+            return val
+        return result + val
+
+    rows = [(1,) for _ in range(10)]
+    numbers = gp.values(rows, db=db, column_names=["val"])
+    results = my_sum(numbers["val"], as_name="result").to_table().fetch()
+    for row in results:
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply tests/test_func.py]: Done.

File: greenplumpython/func.py (hunk 1 above)
Comment (xuebinsu) [reply greenplumpython/func.py]: Done.

File: greenplumpython/func.py [hunk 3]
```diff
@@ -97,6 +130,52 @@ def make_function_call(
...
+    return make_function_call
+
+
+def create_aggregate(
+    trans_func: Callable,
+    name: Optional[str] = None,
+    schema: Optional[str] = None,
+    temp: bool = True,
+    replace_if_exists: bool = False,
+    language_handler: str = "plpython3u",
```
Comment (beeender) [inline greenplumpython/func.py:144]: None of those arguments has test cases. Add FIXME if you plan to add them later.

File: greenplumpython/func.py [hunk 4]
```diff
@@ -97,6 +130,52 @@ def make_function_call(
...
+) -> Callable:
+    @functools.wraps(trans_func)
+    def make_function_call(
+        *args: Expr,
+        group_by: Optional[Iterable[Union[Expr, str]]] = None,
+        as_name: Optional[str] = None,
+        db: Optional[Database] = None,
+    ) -> FunctionCall:
+        trans_func_call = create_function(
+            trans_func, "func_" + uuid4().hex, schema, temp, replace_if_exists, language_handler
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply greenplumpython/func.py:154]: If I create the aggregate with the same name as its transition function, like this:
```
create or replace function my_sum(result int, val int) returns int language plpython3u as $$
if result is None:
    return val
return result + val
$$;

CREATE or replace AGGREGATE my_sum (val int) (
    SFUNC = my_sum,
    STYPE = int
);
```

Then I will get two functions with the same name:
```
gpadmin=# \df my_sum
                            List of functions
 Schema |  Name  | Result data type |     Argument data types     | Type 
--------+--------+------------------+-----------------------------+------
 public | my_sum | integer          | result integer, val integer | func
 public | my_sum | integer          | val integer                 | agg
(2 rows)
```

[the remark this answers is not in the document]
Comment (beeender) [reply greenplumpython/func.py:154]: Use `trans_func.__name__` instead of `func_` to make it easier for debugging?

**Review by beeender — COMMENTED** [review]
A major issue of using `group by` as an aggregate function param is that user is hard to write multiple aggregate functions based on the same `group by` condition.

**Comment by xuebinsu** [comment]
> A major issue of using `group by` as an aggregate function param is that user is hard to write multiple aggregate functions based on the same `group by` condition.

This can be mitigated by CTE fusioning, i.e., merge CTEs whenever possible.

File: tests/test_func.py [hunk 5]
```diff
@@ -91,3 +91,51 @@ def test_func_on_more_than_one_table(db: gp.Database):
...
+    results = list(count(numbers["val"]).to_table().fetch())
+    assert len(results) == 1 and results[0]["count"] == 10
+
+
+def test_agg_group_by(db: gp.Database):
+    rows = [(i, i % 2 == 0) for i in range(10)]
+    numbers = gp.values(rows, db=db, column_names=["val", "is_even"])
+    count = gp.aggregate("count", db=db)
+
+    results = list(count(numbers["val"], group_by=["is_even"]).to_table().fetch())
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply tests/test_func.py:110]: Test added.

File: greenplumpython/func.py (hunk 3 above)
Comment (xuebinsu) [reply greenplumpython/func.py:148]: FIXME added

File: greenplumpython/func.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (xuebinsu) [reply greenplumpython/func.py:158]: The point is, function name cannot be longer than 64 bytes in PostgreSQL. It is risky to append a UUID to the name of the original Python function.

I added a test case for this.

**Review by beeender — APPROVED** [review]

**Pull request merged**